Add support for hashing with sha512 for container V3

stefanberger opened the pull request:
To support NIAP's ( National Information Assurance Partnership) requirement of using sha512 for secure boot rather than sha3-512, add sha512 hashing support for container V3 and its related tools such as create-container.c, print-container.c, crtSignedContainer.sh, and hashkeys.c. Keep the default choice always as sha3-512 so that existing users do not need to change anything. Enable the choice of the hash by supporting a --hash option in the tools and calling a function that takes the choice of hash as an input parameter besides the data to hash and the buffer for the resulting digest.

stefanberger committed:
Fix some indentations by converting spaces to tabs

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
Prepare for support for hashing with sha512

Introduce the defines HASH_ALG_SHA512 and HASH_ALG_SHA3_512 with already
existing values. Use them in places where the raw numbers were used.

Rename sha3_512 to calc_hash and add a first parameter for the caller to
choose which hash to use. Only use HASH_ALG_SHA3_512 in this step replacing
sha3_512 calls with calc_hash calls.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
container: Enable v3 container with sha512 by adding --hash option

Allow users to choose the hash (sha3-512 or sha512) when creating a V3
container. For all other versions only allow the user to choose the
existing version. Adapt the code for V3 only, leave V1 and V2 code to
only support SHA512 and SHA3-512 respectively.

In print-container take the used hash_alg from the V3 header when getting
the payload hash. For all other versions use the hard coded hash algorithm.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
crtSignedContainer: Add support for using sha3 with V3 containers

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
hashkeys: Add support for --hash option usable for container V3

Add support for choosing which hash to use, sha3-512 or sha512,
when choosing a V3 container and pass the choice of the hash to
the calc_hash function.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

cjengel commented on container.h:
```diff
@@ -72,6 +72,9 @@ typedef uint8_t mldsa_signature_t[MLDSA_87_SIG_LENGTH];
+#define HASH_ALG_NONE		0
+#define HASH_ALG_SHA512		1
+#define HASH_ALG_SHA3_512	2
 	uint8_t sig_alg;	/* (1: SHA-512/ECDSA-521) 2: SHA3-512 ECDSA-521/Dilithium r2 8/7
```
Haven't had time to review much of this but did have a comment.  the hash_alg designates the hash used over the payloads and is included in the headers.  Do we also need a new signature algorithm to indicate '4: SHA-512 ECDSA 521/ML-DSA-87'

cmjishnu replied:
I have the same question as Chris. Also should we consider adding it in a new container version?
I understand we'll have to add more code for that.

stefanberger replied:
Yes, there should be a '4: SHA512 ECDSA-521/ML-DSA-87' here.

If container v4 was to be equivalent to signature '4' then we could do away with the --hash option. Either way would work.

cjengel replied:
I would prefer to keep the container version the same (v3) since we don't need to change the structure of the container and just add the new signature algorithhm (4) .  And then the --hash option would influence that along with the hash of the payload

stefanberger replied:
@cjengel I added a new signature algorithm '4' now.

stefanberger committed:
Introduce #defines for values of the sig_alg field

Introduce #defines for the values of the sig_alg field and replace the raw
numbers with the #defines.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

cjengel commented:
`Software Header:
ver_alg:
  version:  0003
  hash_alg: 01 (SHA512)
  sig_alg:  04 (UNKNOWN)
`
print-container is missing parsing of the new sig_alg

cjengel reviewed (commented):
No additional comments, have verified the changes with my local testcases

stefanberger commented:
> `Software Header: ver_alg: version: 0003 hash_alg: 01 (SHA512) sig_alg: 04 (UNKNOWN) ` print-container is missing parsing of the new sig_alg

I fixed this case now.
